src/main/java/org/cloudfoundry/promregator/scanner/ReactiveTargetResolver.java

```diff
@@ -401,6 +422,40 @@ public void setResolvedApplicationId(String resolvedApplicationId) {
 			return itnew;
 		});
 	}
+
+	private Flux<IntermediateTarget> resolveAnnotations(IntermediateTarget it) {
+		if (Boolean.TRUE.equals(it.getConfigTarget().getKubernetesAnnotations())) {
+			Mono<org.cloudfoundry.client.v3.applications.ListApplicationsResponse> response = this.cfAccessor
+				.retrieveAllApplicationsInSpaceV3(it.getResolvedOrgId(), it.getResolvedSpaceId());
```

Comment:
By this way, we are doubling all space-level requests to the CFCC, if the foundation is supporting V3. Or to put it into other words, `resolveApplication` triggers a call to `retrieveAllApplicationIdsInSpace` (V2) and afterwards this logic here will  call `retrieveAllApplicationsInSpaceV3` again. However, mainly these two calls should merely provide the same response - a lot of redundant information is sent over the wire.
To that end I currently wonder whether we shouldn't merge `resolveAnnotations` into `resolveApplication`: If V3 is available, we take `retrieveAllApplicationsInSpaceV3`; if not we fall back to the old `retrieveAllApplicationIdsInSpace`. 

Disclaimer: Didn't check if that was feasible, just thinking out loud here.

Reply:
It's important to note we are doubling space level requests to CFCC from 1 to 2 if you have the annotation filtering enabled yes. I agree this can probably be tweaked to selectively call each endpoint depending on the CAPI version you support. I was concerned of moving core functionality to rely on the V3 APIs though and I knew this PR was already getting pretty big